lib/source_file.cpp

```diff
@@ -218,17 +219,44 @@ static std::string carets(std::string_view source, unsigned start, unsigned end)
     return builder.str();
 }
 
-cstring InputSources::getSourceFragment(const SourceInfo &position, bool useMarker) const {
+cstring InputSources::getSourceFragment(const SourceInfo &position, int trimWidth,
+                                        bool useMarker) const {
     if (!position.isValid()) return ""_cs;
 
     // If the position spans multiple lines, truncate to just the first line
     if (position.getEnd().getLineNumber() > position.getStart().getLineNumber())
-        return getSourceFragment(position.getStart(), useMarker);
+        return getSourceFragment(position.getStart(), trimWidth, useMarker);
 
     std::string_view result = getLine(position.getStart().getLineNumber());
     unsigned int start = position.getStart().getColumnNumber();
     unsigned int end = position.getEnd().getColumnNumber();
 
+    std::string tmp;  // holding place for temp string_view
+    if (trimWidth > 10 && result.size() > (size_t)trimWidth) {
+        if (result.back() == '\n') result.remove_suffix(1);
+        int ltrim = (int)(end + start) / 2 - trimWidth / 2 + 3;
+        if (ltrim > (int)start) ltrim = start;
+        if (ltrim < 4) ltrim = 0;
+        if (ltrim > (int)result.size() - trimWidth + 3) ltrim = result.size() - trimWidth + 3;
+        if (ltrim == 0) {
+            result.remove_suffix(result.size() - trimWidth + 3);
+            tmp = absl::StrCat(result, "...");
```

Comment:
nitpick: I would probably define `const char ELIPSIS[] = "..."; const int ELIPSIS_W = sizeof(ELIPSIS) - 1` and use that instead of `"..."` & `3`.

Reply:
I'm not a huge fan of defining named constants when the name doesn't provide any extra information about the constant.  So while something like `const int DefaultTerminalWidth = 100` would be good, `ELIPSIS` instead of `"..."` doesn't really add anything IMO.